docs: add posts on outgoing hooks integrations (#707)
devopness/devopness: docs/applications/add-application-deploy-status-to-commit.md

   - Bitbucket: [REST APIs: Create a build status for a commit](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-commit-statuses/#api-repositories-workspace-repo-slug-commit-commit-statuses-build-post)
   - Github: [REST API: Create a commit status](https://docs.github.com/en/rest/commits/statuses#create-a-commit-status)
   - Gitlab: [REST API: Set the pipeline status of a commit](https://docs.gitlab.com/ee/api/commits.html#set-the-pipeline-status-of-a-commit)

1. On your local machine, in a terminal window, submit a request to Devopness API endpoint `POST /hooks/outgoing` to create a outgoing for the `action.started` event, replacing `<application_id>`, `<pipeline_id>`, `<target_url>`, `<request_headers>` and `<request_body>`

   - For further instructions, follow the guide [Create an Outgoing Webhook](/docs/webhooks/create-outgoing-webhook)

   {% raw %}
   ```bash
   curl --request POST \
     --url https://api.devopness.com/hooks/outgoing \
     --header 'Accept: application/json' \
     --header 'Authorization: Bearer <access_token>' \
     --header 'Content-Type: application/json' \
     --data '{

>        "name": "GitHub CI(build)",

Reviewer: @thlmenezes regardless of the source provider, will I have to put that name?

Author: I agree, that's me with too much focus on the request body to notice the remaining mentions to Github.

Thanks for your review. I'm on it

Author: Post updated

>        "name": "GitHub CI(build)",

Author: ```suggestion
       "name": "CI(build)",
```

       "action_type": "deploy",
       "resource_type": "application",
       "resource_id": <application_id>,
       "target_url": "https:\/\/<target_url>\/{{ action.triggered_from.hook_parsed_variables.commit_hash }}",
       "settings": {
         "request_headers": [
           {
             "name": "Authorization",
             "value": "Bearer {{ application.source_provider.access_token }}"
           },
           // NOTE: add Request Headers (`<request_headers>`) here
         ],

>          "request_body": {
>            // NOTE: review the fields bellow according to Request Body (`<request_body>`)
>            "state": <"INPROGRESS" | "pending" | "running">,

Reviewer: @thlmenezes I believe that you should not add examples here, I understand that you tried to put the respective 'pending' status of each source provider, but this can cause confusion, for example, the person thinks that he can use any of these for any source provider, so I believe make it better like this:

```suggestion
         "request_body": {
           // NOTE: review the fields bellow according to Request Body of your source provider (`<request_body>`)
           "state": <source_provider_pipeline_status>,
```

Author: Post updated

           "target_url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "description": "Application building started",
           "context": "ci\/devopness(build)",
           "key": "ci\/devopness(build)",
         }
       },
       "trigger_when": {
         "events": [
           "action.started"
         ],
         "conditions": [
           {
             "path": "action.triggered_from.hook_parsed_variables.pipeline_id",
             "accepted_values": [
               <pipeline_id>
             ]
           }
         ]
       }
     }'
   ```
   {% endraw %}

1. Repeat the previous step, changing the `request_body` and `trigger_when.events`, to create the commit status to the `action.failed` state.

   {% raw %}
   ```bash
   …
         "request_body": {
           // NOTE: review the fields bellow according to Request Body (`<request_body>`)

>            "state": <"FAILED" | "failure" | "failed">,

Reviewer: @thlmenezes same comment of

Author: Post updated

           "target_url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "description": "Application building failed",
           "context": "ci\/devopness(build)",
           "key": "ci\/devopness(build)",
         }
       },
       "trigger_when": {
         "events": [
           "action.failed"
         ],
   …
   ```
   {% endraw %}

   - NOTE: the field `request_body.context` needs to be the same for all the action status; this way the same commit status will be updated, instead of creating a new entry for every state.

1. Repeat the previous step, changing the `request_body` and `trigger_when.events`, to create the commit status to the `action.completed` state.

   {% raw %}
   ```bash
   …
         "request_body": {
           // NOTE: review the fields bellow according to Request Body (`<request_body>`)

>            "state": <"SUCCESSFUL" | "success">,

Reviewer: @thlmenezes same comment of

Author: Post updated

---
title: Add Application Deploy Status to Commit
intro: Commit statuses provide a way to tag commits with meta data, like automated build results. Learn how to create outgoing webhooks to update a commit status to match your Devopness application deploy status.
links:
    overview:
    quickstart:
    previous:
    next:
    guides:
    related:
    featured:
---

NOTE: Webhooks, for now, are an API only feature; so this post will guide you through the usage of our API to help you create an outgoing webhook.

NOTE: The examples below use `cURL`, which is a command line utility to be used from a terminal window. If you're not familiar with running commands on a terminal or you simply don't want to use a terminal you could, alternatively, copy the examples and import the `cURL` command line into your favorite API client platform. Here are links with instructions on how to achieve that on some of the most popular API client platforms:

> - Postman: [Importing and exporting data: Importing with cURL commands](https://learning.postman.com/docs/getting-started/importing-and-exporting-data/#importing-with-curl-commands)

Author: Post updated
![image](https://github.com/devopness/devopness/assets/33532177/34490dd6-3900-4dac-aab5-2f538e2b0d17)

> - Postman: [Importing and exporting data: Importing with cURL commands](https://learning.postman.com/docs/getting-started/importing-and-exporting-data/#importing-with-curl-commands)

Author: Links to API clients and Source Providers refactored

- Insomnia: [Import and Export Data: Import Data](https://docs.insomnia.rest/insomnia/import-export-data#import-data)

> NOTE: Insomnia uses curly braces syntax for environment variables, to avoid errors you need to disable the nunjuncks template feature for the request body. Further instructions at [Insomnia's FAQ](https://docs.insomnia.rest/insomnia/faq#how-can-i-temporarily-disable-nunjucks-template)

Author: Post updated
![image](https://github.com/devopness/devopness/assets/33532177/34490dd6-3900-4dac-aab5-2f538e2b0d17)

> NOTE: Insomnia uses curly braces syntax for environment variables, to avoid errors you need to disable the nunjuncks template feature for the request body. Further instructions at [Insomnia's FAQ](https://docs.insomnia.rest/insomnia/faq#how-can-i-temporarily-disable-nunjucks-template)

Reviewer 2: @thlmenezes I reviewed again and no changes related to this conversation. Are you sure you updated the branch?

> NOTE: Insomnia uses curly braces syntax for environment variables, to avoid errors you need to disable the nunjuncks template feature for the request body. Further instructions at [Insomnia's FAQ](https://docs.insomnia.rest/insomnia/faq#how-can-i-temporarily-disable-nunjucks-template)

Author: Sorry @jfoliveira I misinterpret this thread the first time I've read.

I disagree

The note about insomnia is needed to help the user succeed.

The following UX follows the `cURL` request import on insomnia

![error_insomnia](https://github.com/devopness/devopness/assets/33532177/901d96f1-d17a-4e60-9ba4-17aa92f5388c)

And if you try to send the request

![error_insomnia_req](https://github.com/devopness/devopness/assets/33532177/87126891-4bad-4639-9788-4eea32a45bcc)

Many users may use insomnia, since it's a popular tool, and be lost when encountering this error, because the error message and its workaround solution are not obvious; that was the reasoning behind the documentation, to help the user succeed in making the request on all the options cited on the post.

> That's expensive to maintain

I agree, but in this case I believe is needed, since the first moment after importing the data the user is met with an error.

If we remove this note, I suggest removing all Insomnia references

Does this seem like reasonable justification to maintain this note in docs, @jfoliveira ?

> NOTE: Insomnia uses curly braces syntax for environment variables, to avoid errors you need to disable the nunjuncks template feature for the request body. Further instructions at [Insomnia's FAQ](https://docs.insomnia.rest/insomnia/faq#how-can-i-temporarily-disable-nunjucks-template)

Reviewer 2: > The note about insomnia is needed to help the user succeed.

The third-party tool own documentation is not enough to help users, who decided to use that tool, to achieve success on their tool?

> Many users may use insomnia
True. Many MAY, many other MAY NOT. Should our docs care about all possible tools that our users MAY use?
Should we be monitoring market share and every time another tool become popular we need to update our docs?
Can you see that as a sustainable approach?

>     That's expensive to maintain
>> I agree, but in this case I believe is needed,

I'm not convinced, so I left some food for thought for you above and also:

1. Should Devopness team pay the price of documenting edge cases on third-party tools being used by specific users, by their own choice? Are you sure shouldn't the user read the documentation of the tools they decided to use?
2. If the user is not able to read the docs of the external tool they chose, why would they bother reading Devonpess documentation?
3. Likewise, are you also expecting third-party tools to also document Devopness specific behaviors every time we change things on our side?

Please don't take too much time replying to every single topic. If you're convinced, please just summarize the reason why you believe the approach you are suggesting is a common practice in the software industry.

> NOTE: Insomnia uses curly braces syntax for environment variables, to avoid errors you need to disable the nunjuncks template feature for the request body. Further instructions at [Insomnia's FAQ](https://docs.insomnia.rest/insomnia/faq#how-can-i-temporarily-disable-nunjucks-template)

Reviewer 2: @thlmenezes one extra point on your note:

```
**NOTE**: Insomnia uses curly braces syntax for environment variables, to avoid errors you need to 
disable the nunjuncks template feature for the request body. Further instructions at [Insomnia's
```

is bringing a lot of new words and concepts (`curly braces`, `syntax`, `environment variables`, `nunjunkcs`, `template`) to a user who just arrived at this topic with a single goal: create a Devopness outgoing webhook.

Does the average user (which doesn't use or need such specific third-party tool) need to be presented to all of this info?

> NOTE: Insomnia uses curly braces syntax for environment variables, to avoid errors you need to disable the nunjuncks template feature for the request body. Further instructions at [Insomnia's FAQ](https://docs.insomnia.rest/insomnia/faq#how-can-i-temporarily-disable-nunjucks-template)

Author: @jfoliveira removed note from docs.

May the users discover and solve this with the Insomnia team.

     --header 'Accept: application/json' \
     --header 'Authorization: Bearer <access_token>' \
     --header 'Content-Type: application/json' \
     --data '{
       "name": "CI(build)",
       "action_type": "deploy",
       "resource_type": "application",
       "resource_id": <application_id>,
       "target_url": "https:\/\/<target_url>\/{{ action.triggered_from.hook_parsed_variables.commit_hash }}",
       "settings": {
         "request_headers": [
           {
             "name": "Authorization",
             "value": "Bearer {{ application.source_provider.access_token }}"
           }
           // NOTE: add Request Headers (`<request_headers>`) here
         ],
         "request_body": {
           // NOTE: review the fields bellow according to Request Body (`<request_body>`) from the source provider instructions
           "state": <source_provider_pipeline_status>,
           "target_url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "description": "Application building started",
           "context": "ci\/devopness(build)",

>            "key": "ci\/devopness(build)",

Author: ```suggestion
           "key": "ci\/devopness(build)"
```

     --header 'Accept: application/json' \
     --header 'Authorization: Bearer <access_token>' \
     --header 'Content-Type: application/json' \
     --data '{
       "name": "CI(build)",
       "action_type": "deploy",
       "resource_type": "application",
       "resource_id": <application_id>,
       "target_url": "https:\/\/<target_url>\/{{ action.triggered_from.hook_parsed_variables.commit_hash }}",
       "settings": {
         "request_headers": [
           {
             "name": "Authorization",
             "value": "Bearer {{ application.source_provider.access_token }}"
           }
           // NOTE: add Request Headers (`<request_headers>`) here
         ],
         "request_body": {
           // NOTE: review the fields bellow according to Request Body (`<request_body>`) from the source provider instructions
           "state": <source_provider_pipeline_status>,
           "target_url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "description": "Application building failed",
           "context": "ci\/devopness(build)",

>            "key": "ci\/devopness(build)",

Author: ```suggestion
           "key": "ci\/devopness(build)"
```

     --header 'Accept: application/json' \
     --header 'Authorization: Bearer <access_token>' \
     --header 'Content-Type: application/json' \
     --data '{
       "name": "CI(build)",
       "action_type": "deploy",
       "resource_type": "application",
       "resource_id": <application_id>,
       "target_url": "https:\/\/<target_url>\/{{ action.triggered_from.hook_parsed_variables.commit_hash }}",
       "settings": {
         "request_headers": [
           {
             "name": "Authorization",
             "value": "Bearer {{ application.source_provider.access_token }}"
           }
           // NOTE: add Request Headers (`<request_headers>`) here
         ],
         "request_body": {
           // NOTE: review the fields below according to Request Body (`<request_body>`) from the source provider instructions
           "state": <source_provider_pipeline_status>,
           "target_url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "url": "https:\/\/app.devopness.com\/actions\/{{ action.id }}",
           "description": "Application building success",
           "context": "ci\/devopness(build)",

>            "key": "ci\/devopness(build)",

Author: ```suggestion
           "key": "ci\/devopness(build)"
```

intro: Commit statuses provide a way to tag commits with meta data, like automated build results. Learn how to create outgoing webhooks to update a commit status to match your Devopness application deploy status.
links:
    overview:
    quickstart:
    previous:
    next:
    guides:
    related:
    featured:
---

{% note %}

**NOTE**: Webhooks, for now, are an API only feature; so this post will guide you through the usage of our API to help you create an outgoing webhook.

{% endnote %}

{% note %}

**NOTE**: The examples below use `cURL`, which is a command line utility to be used from a terminal window. If you're not familiar with running commands on a terminal or you simply don't want to use a terminal you could, alternatively, copy the examples and import the `cURL` command line into your favorite API client platform. Here are links with instructions on how to achieve that on some of the most popular API client platforms:

- [Postman](https://learning.postman.com/docs/getting-started/importing-and-exporting-data/#importing-with-curl-commands)
- [Insomnia](https://docs.insomnia.rest/insomnia/import-export-data#import-data)

{% endnote %}

> {% note %}
> 
> **NOTE**: Insomnia uses curly braces syntax for environment variables, to avoid errors you need to disable the nunjuncks template feature for the request body. Further instructions at [Insomnia's FAQ](https://docs.insomnia.rest/insomnia/faq#how-can-i-temporarily-disable-nunjucks-template)
> 
> {% endnote %}

Author: ```suggestion
```

1. On your local machine, in a terminal window, submit a request to Devopness API endpoint `POST /hooks/outgoing` to create a outgoing for the `action.completed` event, replacing `<application_id>`, `<pipeline_id>`, `<target_url>`, `<request_headers>` and `<request_body>`

   {% raw %}
   ```bash
   curl --request POST \
     --url https://api.devopness.com/hooks/outgoing \
     --header 'Accept: application/json' \
     --header 'Authorization: Bearer <access_token>' \
     --header 'Content-Type: application/json' \
     --data '{
       "name": "CI(build)",
       "action_type": "deploy",
       "resource_type": "application",
       "resource_id": <application_id>,
       "target_url": "https:\/\/<target_url>\/{{ action.triggered_from.hook_parsed_variables.commit_hash }}",
       "settings": {
         "request_headers": [
           {
             "name": "Authorization",
             "value": "Bearer {{ application.source_provider.access_token }}"
           }
           // NOTE: add Request Headers (`<request_headers>`) here
         ],
         "request_body": {

>            // NOTE: review the fields bellow according to Request Body (`<request_body>`) from the source provider's instructions

Author: ```suggestion
           // NOTE: review the fields below according to Request Body (`<request_body>`) from the source provider instructions
```
